### PR TITLE
feat(cli): add CLI command `idl` to generate IDL from source crate

### PR DESCRIPTION
**Pull request by vobradovich**
Usages
- with current dir
```
cargo sails idl
```

- with `manifest-path`
```
cargo sails idl --manifest-path E:\git\sails-test\app\Cargo.toml
```

### CODE REVIEW

File: rs/cli/src/idlgen.rs [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    cmd.status().context("failed to execute `cargo` command")
+}
+
+enum MetaPathVersion {
+    V1,
+    V2,
+}
+
+pub const META_PATH_V1: &[&str] = &["sails_rs", "meta", "ProgramMeta"];
+pub const META_PATH_V2: &[&str] = &["sails_idl_meta", "ProgramMeta"];
```
Comment (DennisInSky) [inline rs/cli/src/idlgen.rs:286]: Do you want this to be implemented like
```
impl MetaPathVersion {
    path(&self) -> &[str] {
        match *self ...
    }
}
```
and the have consts for MetaPathVersion V1 and V2

**Review by DennisInSky — COMMENTED** [review]
Some CI test would be nice:
- we have one which generates a program
- then you can use generated program for obtaining its idl
- then maybe compare obtained idl with the one generated during program test

File: rs/cli/src/idlgen.rs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+            .find_map(|p| {
+                if p.1.path == META_PATH_V1 {
+                    Some((p.0, MetaPathVersion::V1))
+                } else if p.1.path == META_PATH_V2 {
+                    Some((p.0, MetaPathVersion::V2))
+                } else {
+                    None
+                }
+            })
+            .context("failed to find `sails_rs::meta::ProgramMeta` definition in dependencies")?;
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply rs/cli/src/idlgen.rs:139]: `sails_idl_meta` path is re-exported via `sails_rs::meta` - and should be used like this

File: rs/cli/src/idlgen.rs [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        // find struct implementing `sails_rs::meta::ProgramMeta`
+        let program_struct_path = doc_crate
+            .index
+            .values()
+            .find_map(|idx| is_implement_trait_path(idx, program_meta_id))
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply rs/cli/src/idlgen.rs:144]: this is forbidden by the `program` macro

File: rs/cli/src/idlgen.rs [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+    cmd.env("CARGO_TARGET_DIR", target_dir)
+        .env("__GEAR_WASM_BUILDER_NO_BUILD", "1")
+        .stdout(std::process::Stdio::null()) // Don't pollute output
+        .arg("run")
+        .arg("--manifest-path")
+        .arg(manifest_path.as_str())
+        .arg("--bin")
+        .arg(bin_name);
+    cmd.status().context("failed to execute `cargo` command")
+}
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply rs/cli/src/idlgen.rs:278]: Don't think there's any need for additional buider struct

File: rs/cli/src/idlgen.rs (hunk 1 above)
Comment (vobradovich) [reply rs/cli/src/idlgen.rs:286]: hmm, rafactored

File: rs/cli/src/idlgen.rs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        &self,
+        program_struct_path: &str,
+        meta_path_version: MetaPathVersion,
+    ) -> anyhow::Result<Utf8PathBuf> {
+        // find `sails-rs` dependency
+        let sails_dep = self
+            .program_package
+            .dependencies
+            .iter()
+            .find(|p| p.name == "sails-rs")
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply rs/cli/src/idlgen.rs:165]: It could, but any ideas on how to find out?

File: rs/cli/src/idlgen.rs [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+            .iter()
+            .filter(|&p| p.name == "sails-rs")
+            .collect::<Vec<_>>();
+
+        let target_dir = self
+            .target_dir
+            .as_ref()
+            .unwrap_or(&metadata.target_directory);
+
+        let package_list = get_package_list(&metadata, self.deps_level.unwrap_or(1))?;
```
Comment (DennisInSky) [inline rs/cli/src/idlgen.rs:54]: Do you want to make `deps_level` as usize in the struct and just initialize via `unwrap_or`

File: rs/cli/src/idlgen.rs [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+            package_list.len()
+        );
+        for program_package in package_list {
+            let idl_gen = PackageIdlGenerator::new(
+                program_package,
+                &sails_packages,
+                target_dir,
+                &metadata.workspace_root,
+            );
+            match idl_gen.get_program_struct_path_from_doc() {
```
Comment (DennisInSky) [inline rs/cli/src/idlgen.rs:66]: This method doesn't seem native to ProgramIdlGenerator. I think we wanted to make it as a free func.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -141,9 +141,13 @@ jobs:
...
+          SAILS_CLI_TEMPLATES_BRANCH=${GITHUB_HEAD_REF:-${GITHUB_REF#refs/heads/}} ./target/debug/cargo-sails sails program ~/tmp --name my-demo
 
       - name: Run Tests on MyDemo
         run: |
           cd ~/tmp/my-demo
           cargo test -p my-demo
+
+      - name: Generate IDL from MyDemo via CLI
+        run: |
+          ./target/debug/cargo-sails sails idl --manifest-path ~/tmp/my-demo/Cargo.toml
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply .github/workflows/ci.yml:153]: added

**Review by DennisInSky — APPROVED** [review]

**Pull request merged**